### ISSUE TITLE
RFC: MDX comment syntax

### ISSUE THREAD

**Issue by johno** [mdx-js/mdx#1042]
As part of #1039 comment syntax currently only supports JS comment expressions:

```jsx
# Hello, world!

{/* I am a comment */}
```

This syntax is nice because it feels more JSX-y than HTML comments. However, it comes at a cost of interoperability with Markdown (which uses HTML comments).

I'm inclined to think ensuring that we also support HTML comments (despite it being a bit precarious to have multiple syntaxes) is worth it so that your average Markdown/README file works with MDX out of the box.

```jsx
# Hello, world!

{/* I'm a comment */}
<!-- I'm a comment, too! -->
```

Anyone have any thoughts or opinions here?

**Comment by ChristopherBiscardi** [comment]
I agree

**Comment by wooorm** [comment]
While adding support isn’t very technically complex, I’m relatively strongly against, because it adds mental complexity. MDX already “comes at a cost of interoperability with Markdown”, namely: HTML isn’t supported, and you get JSX in return. I don’t think there should be exceptions to that rule, as those will make it harder to reason about or explain what the language is. JSX is HTML but stricter; being strict is expected.

**Comment by laurieontech** [comment]
In terms of user experience, markdown should work out of the box in my mind. However, as @wooorm rightly points out, it doesn't in certain circumstances. Unless the plan is to support all HTML I'm not sure that an exception for comments has a positive impact. In fact, if we're going for consistency, it makes more sense not to support it.

**Comment by laurieontech** [comment]
I wouldn't break something that currently works in v1 without a need to do so. That being said, the fuzzy line of markdown interoperability is not ideal.

**Comment by jescalan** [comment]
Agreed with the original post, feels like supporting html comments is pretty important, even if actually using them is discouraged.

**Comment by JounQin** [comment]
It seems that would break [remark-lint](https://github.com/remarkjs/remark-lint)'s `disable` comment?

**Comment by wooorm** [comment]
Hmm, indeed.
It would be possible to support comments like this though `{/* lint ignore this */}`

**Comment by leetrout** [comment]
https://github.com/DavidAnson/markdownlint works great for markdown linting and keeping some vanilla markdown files in the same tree as mdx is very useful for my current project.

The remark-based linters don't seem to cover the same rules at markdownlint does and it is a shame to lose the ability to ignore rules like line length via comment syntax (`<!-- markdownlint-disable MD001 MD005 -->`).

**Comment by wooorm** [comment]
@leetrout You’ll run into other problems though: MD !== MDX. markdownlint will miss stuff, and will warn about stuff that isn‘t a problem.

**Comment by jeffreymeng** [comment]
How come in the MDX playground, the comment syntax mentioned here doesn’t seem to work? It seems to be parsed as italics instead.


![F2C52466-95E4-4078-BF11-C0C69F6A577D](https://user-images.githubusercontent.com/15185258/114844555-c53e2300-9d8f-11eb-861d-b7212ef9ce93.jpeg)

**Comment by timiscoding** [comment]
> How come in the MDX playground, the comment syntax mentioned here doesn’t seem to work? It seems to be parsed as italics instead.
Unsure why but if you remove the line break above the comment, it works.

**Comment by kimbaudi** [comment]
the MDX playground is using MDX v1 so only HTML comments will work. JSX comments only work on MDX v2. Also, HTML comments no longer work on v2.  See <https://github.com/mdx-js/mdx/pull/1039>

**Comment by leebyron** [comment]
I was also surprised by this and it posed a real barrier to adopting MDX for me (I have many existing markdown files with comments that I wanted to migrate to MDX and could not). So I wrote a plugin:

https://github.com/leebyron/remark-comment

I think I have an acceptable navigation of the whole "JSX is not HTML" issue:

* HTML comments cannot be included in JSX. Preserves the idea that JSX is just JSX.
* HTML comments are not emitted to the rendered output. This was always something I wanted my markdown renderer to do anyway, so I saw this as a side benefit. However this also side-steps the problem of trying to decide how to emit an HTML comment as a JSX node. It just does not.

I'd love to see this behavior built into MDX proper.

**Comment by wooorm** [comment]
MDX@2 can also parse “normal” markdown. Use the extension `.md` for markdown files and `.mdx` for the MDX files, if you have lots of files and want to slowly migrate.
Alternatively, you can use JS(X) comments instead: `{/* stuff */}`. For a quick and mostly-fine solution, when migrating, you can search for `<!--(.*?)-->` and replace with `{/*$0*/}`.
